Add torch.linalg.vector_norm function

kurtamohler opened the pull request:
Fixes #50214

kurtamohler committed:
Add torch.linalg.vector_norm function

kurtamohler commented:
Hey @mruberry, as we discussed before, it's not strictly necessary to add a `dim` argument to `torch.linalg.vector_norm` in order to replace all the old `torch.norm` calls. However, I think we should it in this PR, for a couple reasons:

- It should be pretty simple to add it, since the code structure of `at::linalg_vector_norm` is very similar to `at::norm`.
- If we add the `dim` argument, we can go ahead and replace the `at::norm` call (plus all the extra overriding bahavior I had to add on top of `at::norm`) in `at::linalg_norm` with `at::linalg_vector_norm` in this PR. This will give us extra assurance that `at::linalg_vector_norm` does exactly what we need it to do.
- Since `dim` is a positional argument, if we don't include it in this PR, we will have to either break backward compatibility to add it in the future or add a second set of entries for it in `native_functions.yaml`. I think it would be nicer to avoid all of that.

Let me know what you think, when you get a chance

kurtamohler commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -218,6 +218,67 @@ Tensor norm_backward(Tensor grad, const Tensor& self, const optional<Scalar> & p
+    // Find the elements from `self` that equal the norm result
+    Tensor is_equal_to_norm;
+
+    // TODO: For some reason, for complex CPU, (self.abs() == norm) does not
```
I need to figure out this issue before this PR can be merged. The workaround below is probably a lot less performant than `self.abs() == norm`, since it's basically redoing the work that the forward call did.

kurtamohler commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1852,6 +1811,89 @@ static Tensor& linalg_norm_out_impl(Tensor& result, const Tensor& self, optional
+  Tensor self_;
+
+  if (self.device().type() == c10::kCPU && std::abs(ord) == INFINITY) {
+    // TODO: For some reason, for complex CPU, at::abs() gives a different
```
I now understand why this happens.

First of all, it's worth mentioning, for clarity, that in order to find out which element from `self` was returned by the `vector_norm` call as the min/max value, the backward method performs `self.abs() == norm_result`. Then, it simply sets the corresponding element(s) in the gradient to 1, and all else to 0.

The root issue that causes the backward function to fail is `self.abs() == norm_result`. `linalg_vector_norm`'s dispatch function `norm_stub` with infinite order is using a per-element `std::abs` call before applying reductions, instead of using `at::abs`. Thus it becomes possible for none of the elements from `self.abs()` to match exactly with `norm_result`. The reason they could mismatch is because `at::abs` is free to use a vectorized abs function, which can produce slightly different results than `std::abs`.

Below is an example of a situation where vectorized and nonvectorized `torch.abs` give different results with a complex input (at least in my environment):

```
import torch
b = torch.tensor([(7.652771299449274e-06-1.7120581036111606e-06j), (1.9727849811062448e-05+1.919867619816995e-05j), (1.6650342521593928e-05-1.0042102916947193e-05j), (-4.901525595762313e-06-6.710164915302073e-06j)], dtype=torch.complex128)
print(b.abs().tolist())
```

Run with `ATEN_CPU_CAPABILITY=default` to disable vectorization, it prints:
```
[7.841941820226363e-06, 2.7527753739272188e-05, 1.9444221174450732e-05, 8.309709161966138e-06]
```

Run with `ATEN_CPU_CAPABILITY=avx2` to enable vectorization, it prints:
```
[7.841941820226363e-06, 2.752775373927219e-05, 1.944422117445073e-05, 8.309709161966138e-06]
```

Notice that the second and third values in both outputs are slightly different. The backward `at::abs()` call is free to be vectorized, so it can disagree with the `std::abs` per-element call within `norm_stub`.

The workaround that I added here simply calls `at::abs` if an infinite order was given to the CPU implementation, so it avoids the complex per-element `std::abs` call in `norm_stub`. The forward `at::abs` call should give the exact same result as the backward `at::abs` call. But that is not ideal since dispatching to `at::abs()` has some extra overhead that I would rather avoid if possible.

Furthermore, the strategy that `linalg_vector_norm_backward` uses to determine which element was chosen from the input is also not ideal. It takes extra time to recalculate information (`self.abs()`) that could potentially be captured by the forward function. But as far as understand, in order to do that we would need to add an extra return value (in a `std::tuple`) to `linalg_vector_norm`. Or is there some other way to do it? I'm going to look into how `torch.amax/amin` do this--perhaps they use a good strategy that I can mimic.

kurtamohler replied:
@ngimel or @mruberry, do have some insight to what I should do about this problem? Apologies if my description is unclear, let me know if so.

ngimel replied:
How's current `norm` dealing with this problem?

kurtamohler replied:
It doesn't. `linalg_vector_norm_backward` and `norm_backward` use the same technique for infinite orders. If I add an OpInfo for `torch.norm`, it fails in the same way that `torch.linalg.vector_norm` fails. `torch.linalg.norm`, on the other hand, passes in this case in the current main branch, because it's directly calling `at::abs` in the forward: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/LinearAlgebra.cpp#L1694

kurtamohler replied:
A possible solution could be to vectorize the CPU implementation of `vector_norm`. As long as vectorization for the absolute value calculation is applied in the same way that `at::abs()` does it, then the backward and forward calls would match. This should also improve performance a fair bit.

mruberry replied:
cc @albanD to also weigh in on this issue, and I've added it to our sync discussion for tomorrow as well

mruberry replied:
@albanD @ngimel and I had an opportunity to discuss and we'd like to propose keeping the working of unfused abs in forward as long as it's limited to cpu complex and inf order. Would you file a follow-up issue, however, @kurtamohler, describing this challenge? We would like to consider more significant changes to improve the performance here in the future.

kurtamohler replied:
Filed #52648

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -422,6 +422,64 @@ def sample_inputs_linalg_norm(op_info, device, dtype, requires_grad):
+                if is_dtype_half and ord in [inf, -inf]:
+                    continue
+
+                for dim in dim_args:
```
I like how thorough this test is and appreciate its structure, but I think this only needs to append one set of these SampleInputs? I also think this can enumerate the product of keepdim, dim_args, and ords to avoid nesting.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -378,6 +378,73 @@
+
+Args:
+    input (Tensor): The input tensor. If the input has more than one dimension,
+        it is interpretted as a tensor flattened to one dimension.  The input
```
"The input data type must be either a floating point or complex type." 

This is good but we should try to consistently use "Supports..." statements like torch.sub() does (see https://pytorch.org/docs/master/generated/torch.sub.html?highlight=sub#torch.sub) and enumerate the supported dtypes there at the end of the description.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -378,6 +378,73 @@
+    input (Tensor): The input tensor. If the input has more than one dimension,
+        it is interpretted as a tensor flattened to one dimension.  The input
+        data type must be either a floating point or complex type. For complex
+        inputs, the norm is calculated on of the absolute values of each
```
"is calculated on of the absolute values of each element" -> "is calculated using the absolute value of each element"

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -378,6 +378,73 @@
+        corresponding floating point type (e.g. float if :attr:`input` is
+        complex float).
+
+    ord (int, float, inf, -inf, optional): The order of norm.
```
"The order of norm" -> "The order of the norm."

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -378,6 +378,73 @@
+        complex float).
+
+    ord (int, float, inf, -inf, optional): The order of norm.
+        inf refers to :attr:`float('inf')`, numpy's :attr:`inf` object, or any
```
I think we can strike this sentence. "inf" seems clear. If we do reference NumPy the docs should capitalize it.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -378,6 +378,73 @@
+        inf    max(abs(x))
+        -inf   min(abs(x))
+        0      sum(x != 0)
+        1      as below
```
Why separate these from the general case?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -378,6 +378,73 @@
 """)
 
+vector_norm = _add_docstr(_linalg.linalg_vector_norm, r"""
+linalg.vector_norm(input, ord=None, keepdim=False, *, out=None, dtype=None) -> Tensor
```
out= is traditionally the last parameter

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -378,6 +378,73 @@
+linalg.vector_norm(input, ord=None, keepdim=False, *, out=None, dtype=None) -> Tensor
+
+Returns the vector norm of a given tensor. 
+
```
What about a reference to torch.linalg.norm like, "See torch.linalg.norm for computing matrix norms"?

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -378,6 +378,73 @@
+
+    out (Tensor, optional): The output tensor. Ignored if ``None``. Default: ``None``
+
+    dtype (:class:`torch.dtype`, optional): If specified, the input tensor is cast to
```
"the input tensor" -> ":attr:\`input\`"

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -378,6 +378,73 @@
+
+    dtype (:class:`torch.dtype`, optional): If specified, the input tensor is cast to
+        :attr:`dtype` before performing the operation, and the returned tensor's type
+        will be :attr:`dtype`. If this argument is used in conjunction with the
```
This limitation is OK for now but in the future we should support safe casting.

cc @IvanYashchuk who just fixed a swathe of linalg functions. Depending on when those PRs land it may be convenient to update this op's behavior in this PR, too.

kurtamohler replied:
Just for my own reference, I believe this is the PR you're talking about: #51560

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -378,6 +378,73 @@
             [1, 2, 2, 2]])
 """)
 
+vector_norm = _add_docstr(_linalg.linalg_vector_norm, r"""
```
This function needs to be added to linalg.rst to appear in the docs.

mruberry commented on test/test_linalg.py:
```diff
@@ -1238,6 +1379,26 @@ def gen_error_message(input_size, ord, keepdim, dim=None):
+        for pair in itertools.product([inf, -inf, 0.0, nan, 1.0], repeat=2):
+            vectors.append(list(pair))
+        for vector in vectors:
+            x = torch.tensor(vector).to(device)
```
`tensor(vector, device=device)`

mruberry commented on test/test_linalg.py:
```diff
@@ -963,6 +963,147 @@ def get_value_type(dtype):
+                with self.assertRaises(RuntimeError):
+                    torch.linalg.vector_norm(input, ord, keepdim=keepdim)
+            else:
+                result_preflattened = torch.linalg.norm(input_flat, ord, keepdim=keepdim)
```
This is good but I suggest making a reference function, like vector_norm_reference(), with the same signature as vector norm except it uses this alternative (Python) implementation. Then these tests could just call that reference and compare the results.

mruberry commented on test/test_linalg.py:
```diff
@@ -963,6 +963,147 @@ def get_value_type(dtype):
+            ((3, 4, 5, 6, 7), []),
+
+            # Infinite orders cause an error for degenerate inputs because
+            # min/max cannot be called on degernate inputs
```
"degernate" -> "degenerate"

mruberry commented on test/test_linalg.py:
```diff
@@ -963,6 +963,147 @@ def get_value_type(dtype):
+
+        for keepdim in [True, False]:
+            for input_size, error_ords in test_cases:
+                input = torch.randn(*input_size, dtype=dtype, device=device)
```
Use make_tensor instead of randn

mruberry commented on test/test_linalg.py:
```diff
@@ -963,6 +963,147 @@ def get_value_type(dtype):
+                result = torch.linalg.vector_norm(input, ord, keepdim=keepdim)
+                self.assertEqual(result, result_preflattened, msg=msg)
+
+        for keepdim in [True, False]:
```
This can just enumerate the product of these options and not worry about detaching and cloning the inputs

mruberry commented on test/test_linalg.py:
```diff
@@ -963,6 +963,147 @@ def get_value_type(dtype):
+
+        def run_test_case(input, ord, keepdim, should_error):
+            msg = f'input.size()={input.size()}, ord={ord}, keepdim={keepdim}, dtype={dtype}'
+            input_copy = input.clone().detach()
```
I don't think you need to clone or detach here?

mruberry commented on test/test_linalg.py:
```diff
@@ -963,6 +963,147 @@ def get_value_type(dtype):
 
+    # This test confirms that torch.linalg.vector_norm's dtype argument works
+    # as expected, according to the function's documentation
+    @skipCUDAIfNoMagma
```
This is a very thorough test. If a reference is created (as suggested below) then could that later test be expanded by also enumerating over dtype and just validating that the behavior is the same as linalg.norm's?

kurtamohler replied:
Ah yeah, that's a really good simplification

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -9006,6 +9006,19 @@
+    DefaultBackend: linalg_vector_norm
+
+- func: linalg_vector_norm.out(Tensor self, Scalar? ord=None, int[1]? dim=None, bool keepdim=False, *, ScalarType? dtype=None, Tensor(a!) out) -> Tensor(a!)
+  use_c10_dispatcher: hacky_wrapper_for_legacy_signatures
```
Can this ditch the hacky wrapper?

kurtamohler replied:
Looks like every `linalg_*.out` function except for `linalg_slogdet.out` uses this. If I remove the wrapper, I get a linker error. I'm not sure what's different about `linalg_slogdet.out`, I'll see if I can find out.

kurtamohler replied:
I figured it out. Evidently, if I don't use `hacky_wrapper_for_legacy_signatures`, the `at::*_out` function needs to be declared with `result` as the last argument: https://github.com/pytorch/pytorch/blob/a51b9a823cff20d5a0bcd299e11e5cf542814640/aten/src/ATen/native/README.md#use_c10_dispatcher

kurtamohler replied:
I suppose all linalg functions should be updated to avoid `hacky_wrapper_for_legacy_signatures` eventually, right?

mruberry commented on aten/src/ATen/native/cuda/ReduceNormKernel.cu:
```diff
@@ -28,9 +28,9 @@ void norm_kernel_cuda_impl(TensorIterator& iter, Scalar val) {
     gpu_reduce_kernel<scalar_t, out_t>(iter, NormTwoOps<scalar_t, acc_t>(), 0);
   } else if (p == static_cast<double>(INFINITY)) {
-    gpu_reduce_kernel<scalar_t, out_t>(iter, AbsMaxOps<scalar_t, acc_t>(), 0);
+    gpu_reduce_kernel<scalar_t, out_t>(iter, AbsMaxOpsOld<scalar_t, acc_t>(), 0);
```
Would you elaborate on this change?

mruberry commented on aten/src/ATen/native/cuda/LinearAlgebra.cu:
```diff
@@ -492,8 +495,70 @@ void addr_kernel_cuda(TensorIterator &iter, Scalar beta, Scalar alpha) {
+  } else if (val.isFloatingPoint()) {
+     ord_val = val.to<double>();
+  } else {
+     AT_ERROR("linalg_vector_norm_kernel_cuda_impl expects norm to be integer or float");
```
`TORCH_CHECK(false, ...)`

AT_ERROR() is deprecated

mruberry commented on aten/src/ATen/native/cuda/LinearAlgebra.cu:
```diff
@@ -492,8 +495,70 @@ void addr_kernel_cuda(TensorIterator &iter, Scalar beta, Scalar alpha) {
+      gpu_reduce_kernel<scalar_t, out_t>(iter, NormZeroOps<scalar_t, acc_t>(), 0);
+    } else if (ord_val == static_cast<double>(1)) {
+      gpu_reduce_kernel<scalar_t, out_t>(iter, NormOneOps<scalar_t, acc_t>(), 0);
+    } else if (ord_val == static_cast<double>(2)) {
```
Are these static casts necessary?

kurtamohler replied:
Evidently not, I'll remove them

mruberry commented on aten/src/ATen/native/cuda/LinearAlgebra.cu:
```diff
@@ -492,8 +495,70 @@ void addr_kernel_cuda(TensorIterator &iter, Scalar beta, Scalar alpha) {
+    } else if (ord_val == static_cast<double>(-INFINITY)) {
+      gpu_reduce_kernel<scalar_t, out_t>(iter, AbsMinOps<scalar_t, acc_t>(), std::numeric_limits<acc_t>::infinity());
+    } else {
+      gpu_reduce_kernel<scalar_t, out_t>(iter, NormOps<scalar_t, acc_t>{ static_cast<acc_t>(ord_val) }, 0);
```
What happens if a large integer value is given that's not representable as a double?

kurtamohler replied:
That's an interesting point. If ord is a large enough integer value, converting it to a double will introduce some error:
```
>>> ord = torch.tensor(2**62 - 10, dtype=torch.long)
>>> ord.to(torch.float64).to(torch.long) - ord
tensor(10)
```

However, with an ord value this large, I think the norm result will almost always be either infinity (if abs value of any input element is slightly greater than 1), 1 (if abs value of at least one element is exactly 1, and all are less than or equal to 1), or 0 (if abs value of all elements is less than 1). At the moment, I don't see any other possible result.

```
>>> ord = torch.tensor(2**62 - 10, dtype=torch.long)
>>> torch.linalg.norm(torch.tensor([1, 0.1, 0, 1.00000001], dtype=torch.double), ord)
tensor(inf, dtype=torch.float64)
>>> torch.linalg.norm(torch.tensor([1, 0.1, 0, 0], dtype=torch.double), ord)
tensor(1., dtype=torch.float64)
>>> torch.linalg.norm(torch.tensor([0.23, 0.1, 0.999999999999999, 0], dtype=torch.double), ord)
tensor(0., dtype=torch.float64)
```

So I'm inclined to suggest that the error introduced when using a very large integer ord value probably won't have any practical consequences. What do you think?

mruberry commented on aten/src/ATen/native/cuda/LinearAlgebra.cu:
```diff
@@ -492,8 +495,70 @@ void addr_kernel_cuda(TensorIterator &iter, Scalar beta, Scalar alpha) {
+template <typename scalar_t, typename acc_t=typename scalar_value_type<scalar_t>::type, typename out_t=typename scalar_value_type<scalar_t>::type>
+void linalg_vector_norm_kernel_cuda_impl(TensorIterator& iter, Scalar val) {
+  double ord_val;
+  if (val.isIntegral(false)) {
```
Add the /*is_boolean=*/ (or whatever the exact word is) comment for readability

mruberry commented on aten/src/ATen/native/cuda/LinearAlgebra.cu:
```diff
@@ -492,8 +495,70 @@ void addr_kernel_cuda(TensorIterator &iter, Scalar beta, Scalar alpha) {
+    } else {
+      gpu_reduce_kernel<scalar_t, out_t>(iter, NormOps<scalar_t, acc_t>{ static_cast<acc_t>(ord_val) }, 0);
+    }
+    // For complex outputs, the above kernels do not touch the imaginary values,
```
This is interesting. If a double is written to a complex value does it not zero the imaginary part?

kurtamohler replied:
Yeah, unfortunately, if I remove this check, the imaginary part doesn't get initialized, so it's filled with garbage:

```
>>> a = torch.randn(10, device='cuda:0', dtype=torch.float32)
>>> torch.linalg.vector_norm(a, dtype=torch.complex64)
tensor(3.4786+2.5625j, device='cuda:0')
>>> torch.linalg.vector_norm(a, dtype=torch.complex64)
tensor(3.4786+2.0703j, device='cuda:0')
```

This happens with all combinations of input types and complex output types that I've tried.

mruberry commented on aten/src/ATen/native/cuda/LinearAlgebra.cu:
```diff
@@ -492,8 +495,70 @@ void addr_kernel_cuda(TensorIterator &iter, Scalar beta, Scalar alpha) {
+  } else {
+     AT_ERROR("linalg_vector_norm_kernel_cuda_impl expects norm to be integer or float");
+  }
+  if (iter.numel() != 0) {
```
This might be more readable with the opposite check:

```
if (iter.numel() == 0) {
  iter.output().fill_((ord_val < 0) ? INFINITY : 0);
  return;
}
```

Then the remaining code doesn't need to be nested and the final `else` clause in this function isn't so far separated from its conditional logic.

mruberry commented on aten/src/ATen/native/cpu/LinearAlgebraKernel.cpp:
```diff
@@ -79,8 +81,50 @@ void addr_kernel(TensorIterator &iter,
 }
 
+template <typename scalar_t, typename acc_t=typename scalar_value_type<scalar_t>::type>
+void linalg_vector_norm_kernel_cpu_impl(TensorIterator& iter, Scalar val) {
```
Same comments here as for the CUDA version.

mruberry commented on aten/src/ATen/native/SharedReduceOps.h:
```diff
@@ -189,12 +192,53 @@ struct AbsMinOps {
 // a set of numbers.
 // `scalar_t` is the type of the input and `acc_t` is the type of the accumulated
 // value. These types differ for complex number input support.
+// TODO: Once `torch.norm` is removed and replaced with `torch.linalg.norm`,
```
Add the reason this is "Old." Would "AbsMaxOpsDeprecated" be more accurate? 

Instead of deprecating this, however, is there a world where we want AbsMaxOpsIgnoreNaN? NumPy has a lot of "nan-" reductions, and SciPy has arguments for controlling how NaNs are treated.

kurtamohler replied:
I like the AbsMaxOpsIgnoreNaN suggestion, there could be a plausible reason to keep using this, even if `torch.norm` is removed

ngimel replied:
Incorrect nan behavior currently is due to bug in MIN/MAX implementation https://github.com/pytorch/pytorch/blob/b2d8f0a431f8ce2c606ed25ac4d28287489c71bc/aten/src/ATen/native/SharedReduceOps.h#L23-L29 on cuda only. We should fix it and have single version of AbsMaxOps.

mruberry replied:
I suppose we can always add back AbsMaxOpsIgnoreNaN later if we need it ;)

I spoke to @ngimel offline, @kurtamohler, and if it's OK with you we would like to fix the NaN propagation behavior in this PR. We'll need to add a test for that change for torch.norm, too. For the complex backward issue, that affects torch.norm, too, right? Would you file an issue for that incorrect torch.norm behavior? Whether we address that issue may depend on whether we actually deprecate torch.norm before the 1.9 release or not.

kurtamohler replied:
Got it, sounds good. Yeah I believe torch.norm's complex backward would be affected, but I'll double check first

mruberry commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -664,7 +664,7 @@ static Tensor& norm_out(Tensor &result, const Tensor &self, optional<Scalar> opt
   auto p = opt_p.value_or(2.0).to<double>();
   TORCH_CHECK(self.device().type() == DeviceType::CPU || self.device().type() == DeviceType::CUDA,
-              "norm only supports CPU AND CUDA device type, got: ", self.device().type());
+              "norm only supports CPU and CUDA device type, got: ", self.device().type());
```
"device type" -> "device types"

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1841,7 +1798,9 @@ static Tensor& linalg_norm_out_impl(Tensor& result, const Tensor& self, optional
     std::vector<int64_t> dim_ = opt_dim.has_value() ? opt_dim.value().vec() : make_dim_list(ndim);
     if (!opt_num_ord.has_value() || dim_.size() == 1) {
-      _linalg_norm_vector_out(result, self, opt_num_ord, dim_, keepdim, opt_dtype);
+      Tensor result_ = at::linalg_vector_norm(self, opt_num_ord, opt_dim, keepdim, opt_dtype);
```
Why doesn't this call linalg_vector_norm_out?

kurtamohler replied:
`linalg_vector_norm_out` is not differentiable. In a future PR, I would like to change it to use `linalg_vector_norm_out` only when the out variant is called, and `linalg_vector_norm` otherwise

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1852,6 +1811,89 @@ static Tensor& linalg_norm_out_impl(Tensor& result, const Tensor& self, optional
 }
 
+static Tensor& linalg_vector_norm_impl(Tensor& result, const Tensor& self, optional<Scalar> opt_ord, optional<IntArrayRef> opt_dim, bool keepdim, optional<ScalarType> opt_dtype) {
+  auto ord = opt_ord.value_or(2).toDouble();
```
What happens if ord is a large integer value not exactly representable as a double?

mruberry replied:
Per offline discussion, let's just add a comment

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1852,6 +1811,89 @@ static Tensor& linalg_norm_out_impl(Tensor& result, const Tensor& self, optional
+  auto ord = opt_ord.value_or(2).toDouble();
+
+  TORCH_CHECK(self.device().type() == DeviceType::CPU || self.device().type() == DeviceType::CUDA,
+              "norm only supports CPU AND CUDA device type, got: ", self.device().type());
```
"CPU and CUDA device types"

mruberry replied:
", got:" -> ", but got:"

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1852,6 +1811,89 @@ static Tensor& linalg_norm_out_impl(Tensor& result, const Tensor& self, optional
+  auto ord = opt_ord.value_or(2).toDouble();
+
+  TORCH_CHECK(self.device().type() == DeviceType::CPU || self.device().type() == DeviceType::CUDA,
+              "norm only supports CPU AND CUDA device type, got: ", self.device().type());
```
"norm" -> "vector_norm"

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1852,6 +1811,89 @@ static Tensor& linalg_norm_out_impl(Tensor& result, const Tensor& self, optional
+  TORCH_CHECK(self.device().type() == DeviceType::CPU || self.device().type() == DeviceType::CUDA,
+              "norm only supports CPU AND CUDA device type, got: ", self.device().type());
+  TORCH_CHECK(self.layout() == Layout::Strided,
+              "norm only supports strided layout, got: ", self.layout());
```
"norm" -> "vector_norm"
", got" -> ", but got"

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1852,6 +1811,89 @@ static Tensor& linalg_norm_out_impl(Tensor& result, const Tensor& self, optional
+  ScalarType in_dtype = opt_dtype.has_value() ? opt_dtype.value() : self.scalar_type();
+  TORCH_CHECK(
+      at::isFloatingType(in_dtype) || at::isComplexType(in_dtype),
+      "Can only calculate the norm of floating point and complex dtypes. Got ",
```
To be consistent with the above, I think this would be:

"vector_norm only supports floating point and complex dtypes, but got..."

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1852,6 +1811,89 @@ static Tensor& linalg_norm_out_impl(Tensor& result, const Tensor& self, optional
+      at::isFloatingType(in_dtype) || at::isComplexType(in_dtype),
+      "Can only calculate the norm of floating point and complex dtypes. Got ",
+      toString(in_dtype),
+      " instead.");
```
This can be cut.

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1852,6 +1811,89 @@ static Tensor& linalg_norm_out_impl(Tensor& result, const Tensor& self, optional
+      toString(in_dtype),
+      " instead.");
+
+  IntArrayRef dim = opt_dim.value_or(IntArrayRef{});
```
Where are these wrapped?

kurtamohler replied:
`make_reduction` and `size` are the only calls below that use values from the `dim`, and they have wrapping built in

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1852,6 +1811,89 @@ static Tensor& linalg_norm_out_impl(Tensor& result, const Tensor& self, optional
+    }
+    if (!has_identity) {
+      TORCH_CHECK(ord != INFINITY, (
+        "cannot perform reduction function max on tensor with "
```
"vector_norm cannot..."

The reference to the "reduction function max" is inconsistent with the user's input, which is ord=inf. Maybe:

"vector_norm cannot compute the infinity norm on an empty tensor because the operation does not have an identity"?

kurtamohler replied:
Sounds good to me

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1852,6 +1811,89 @@ static Tensor& linalg_norm_out_impl(Tensor& result, const Tensor& self, optional
+      TORCH_CHECK(ord != INFINITY, (
+        "cannot perform reduction function max on tensor with "
+        "no elements because the operation does not have an identity"));
+      TORCH_CHECK(ord != -INFINITY, (
```
Analogous comments here

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1852,6 +1811,89 @@ static Tensor& linalg_norm_out_impl(Tensor& result, const Tensor& self, optional
+
+  linalg_vector_norm_stub(iter.device_type(), iter, ord);
+
+  if (result_.dtype() != out_dtype) {
```
I'm confused about what's going on here -- what if result is undefined?

In particular, what's the problem with just passing result directly to make_reduction? Is it that reductions don't support complex -> real?

kurtamohler replied:
You're right, there is a simpler way to do this. `result_.dtype() != out_dtype` is always false anyway, so this was a mistake.

I have a better implementation on my local branch, but I haven't pushed it yet because I've been benchmarking this function and speeding things up to at least match what we get with `torch.norm`. Once that's done, I'll push my updates and make a post about performance

mruberry reviewed (commented):
Hey @kurtamohler!

This is awesome! And it's exciting to be exploring new territory with a new linear algebra function!

I'm sorry I wasn't able to review this PR before now, by the way. It's been a busy time. Worse, both @ngimel and I will be on PTO next week (there's a holiday in the US next week so it's a popular vacation time)! So we likely won't be able to get back to this for another week! Sorry about that.

I added some comments and questions inline and am eager to hear your thoughts on them. I'll sync with @albanD tomorrow about the autograd issues, too. I appreciate your in-depth analysis. It's incredibly interesting.

kurtamohler commented:
Thanks for the detailed review Mike! I've addressed most comments, but not all. I'll ping when it's ready

kurtamohler commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -419,6 +419,52 @@ def sample_inputs_linalg_norm(op_info, device, dtype, requires_grad):
+        all_dims = list(range(ndim))
+        for num_reduction_dims in range(1, len(test_size) + 1):
+            for dim in combinations(all_dims, num_reduction_dims):
+                dim_args.add(dim)
```
I'm getting mypy failures here:
```
Traceback (most recent call last):
  File "test_type_hints.py", line 173, in test_run_mypy
    self.fail(f"mypy failed: {stdout} {stderr}")
AssertionError: mypy failed: torch/testing/_internal/common_methods_invocations.py:445: error: Argument 1 to "add" of "set" has incompatible type "Tuple[int, ...]"; expected "None"  [arg-type]
```

I've tried several different ways of doing this, including:
* Make `dim_args` a list
* Make `dim_args` a set
* Initialize `dim_args` as empty, and append/add `None` later
* Initialize `dim_args` including `None`

None of these avoid the mypy error. I'll keep trying to figure out the right way to do it, but if someone knows what to do, please let me know

kurtamohler replied:
Looks like I just needed to define `dim_args` as `dim_args: List[Optional[Tuple[int, ...]]] = [None]`

kurtamohler commented on test/test_linalg.py:
```diff
@@ -1277,9 +1329,39 @@ def gen_error_message(input_size, ord, keepdim, dim=None):
+
     # Test that linal.norm gives the same result as numpy when inputs
     # contain extreme values (inf, -inf, nan)
-    @skipCUDAIf(True, r"GPU Test is blocking torch.svd https://github.com/pytorch/pytorch/pull/48436")
```
This was skipping more than it needs to. `torch.svd` is only called in matrix +/-2 and "nuc". I wanted to add back the unnecessarily skipped cases in this PR to make sure my changes to `linalg.norm` didn't break its extreme value behavior for vector norms.

kurtamohler commented:
I'm ready to report on the performance with my most recent updates to this PR.

**Summary:**

At the moment, this PR does not seem to negatively impact `torch.linalg.norm`, and `torch.linalg.vector_norm` seems to be a performant replacement for the old `torch.norm` function for almost every case I tried except for CPU complex infinite orders with `ATEN_CPU_CAPABILITY=default` (CPU complex infinite orders with AVX2 actually gets a speedup compared to `torch.norm`, because the `abs` call in the forward is vectorized). So, in my opinion, the performance seems to be good enough for this PR.

@mruberry and @ngimel, I'd like to hear your thoughts on this. Please let me know if anything is unclear or if you see something I overlooked.

**Details:**

<details>
<summary>Click to expand</summary>

I've benchmarked much of `torch.linalg.vector_norm`, `torch.linalg.norm`, and `torch.norm`'s functionality with this script: https://github.com/kurtamohler/pytorch-perf-test-scripts/blob/master/linalg_norm/norm_perf.py

I've only run CUDA and single core CPU, no multicore. I used taskset to pin to one core, using this: https://github.com/kurtamohler/pytorch-perf-test-scripts/blob/master/linalg_norm/run_single_core_norm_perf.sh

I also used both `ATEN_CPU_CAPABILITY=default` and `ATEN_CPU_CAPABILITY=avx2` to measure the difference between disabling and enabling AVX2 vectorization.

Some of the methodology I used in this benchmark script:

- For each test case, run with 10 different randomly generated inputs, and pick the median measured time (to decrease uncontrollable variance due to things like OS context switching and frequency changes)
- Run every timed loop twice, and only use the measurement for the second loop. The first loop is warmup
- Number of timed iterations varies based on input size and compute device. I hand-tuned this so that the total time of every timed loop is roughly the same order of magnitude for my hardware. This increases measurement stability. Perhaps it would be better to automatically adjust iteration count, but I didn't do that
- If using CUDA, sync device before stopping the timer
- Compare results of each norm function using the same input, to ensure that they really are doing the same thing

I compared the single call time for `torch.norm`, `torch.linalg.norm`, `torch.linalg.vector_norm`, and a basic Python implementation of the norm function. My measurements are in this spreadsheet: https://docs.google.com/spreadsheets/d/1m0cD14KuTz83bI9WiTMwRFYqbRghw3IuwZ9VBqckodk/edit?usp=sharing

Each tab is layed out the same way. The different tabs show time measurements using different order values (0.5, 1, 2, 3.5, inf, -inf) and ATEN_CPU_CAPABILITY settings ("default" and "avx2"). The raw call time measurements are hidden in columns F through I, because I think it's easier to understand what's going on by looking at speedup comparisons. So I made columns for the following speedups: linalg_norm over norm (column K), linalg_vector_norm over norm (column L), and linalg_norm over the reference python implementation (column N).

**Comparison with `torch.norm`**

For the two columns (K and L) that compare speedup of `linalg` funcs over torch.norm, speedups of 1 or greater are good, indicating that the `linalg` norm functions either match or improve upon `torch.norm`. Anything significantly lower than 1 is highlighted red, showing that the `linalg` function is not as fast as `torch.norm` for that case.

There are only a couple cases where `linalg.vector_norm` is not as fast as `torch.norm`: when the input is a CPU complex tensor and the norm order is +/- inf. Interestingly, if AVX2 is turned on, we actually get a speedup as long as the input is not tiny. The worst speedups we get for `linalg.vector_norm` are roughly 78 - 80%, which is not terrible. This performance decrease is due to the `at::abs()` call that I had to add in `vector_norm`'s forward function to match `at::abs()` call in the backward function. The reason why turning AVX2 on makes `vector_norm` faster than `torch.norm` in this case is that the `at::abs()` call is vectorized. Tiny inputs don't get this performance boost, but that doesn't seem extremely concerning. This performance decrease certainly should be dealt with in a future PR, but it doesn't seem bad enough to delay this PR IMO.

`linalg.norm` shares the same aforementioned performance decrease compared to `torch.norm`, since it calls `at::linalg_vector_norm` internally. It also has one additional problem: All of the CUDA calls are slowed down because of the result tensor resize and copy, due to the fact that `linalg.norm` does not use the out-variant of `linalg.vector_norm` internally. Evidently, the resize and copy doesn't affect CPU impl significantly. I would like to update to call the out-variant, but since out variants are not differentiable, it will require a fair amount of code reorganization within `linalg.norm`. Also, `linalg.norm` already uses a resize and copy on the master branch, so this slowdown already existed. So I think this issue should be solved in a future PR that reorganizes `linalg_norm` so that it can use out-variants where possible (matrix norms also use the less performant resize and copy).

**Comparison with Python reference implementation**

For the last column (N), speedups less than 1 are good, because they indicate that `torch.linalg.norm` is faster than the naive Python implementation. Ideally, everything in this column would be green, but there are lots of cases where the Python impl is faster. There are some slowdowns that I don't understand yet (for instance, why is order 0.5 bad for CPU float32 with AVX2 turned off, while order 3.5 is good in this case?), and some cases can be explained with vectorization. In any case, I don't think this comparison should hold up this PR, but it should be addressed in future PR(s) that bring `linalg.norm` and `linalg.vector_norm` closer to their ideal performance.

</details>

kurtamohler commented on aten/src/ATen/native/SharedReduceOps.h:
```diff
@@ -21,9 +21,26 @@
+    return ::min(a, b);
+  }
+}
+#define MAX(X, Y) max_propagate_nan(X,Y)
```
I'm not sure if this is the best way to add nan propagation. Maybe there's something more standard?

mruberry replied:
@ngimel and I like this approach - is there an updated test for torch.norm nan propagation?

kurtamohler replied:
Yep, I added one

kurtamohler commented:
I think this is ready for another look @mruberry

kurtamohler commented:
Just FYI, the statuses of three CI jobs below are still "pending", but if you click on details for them, they actually finished and passed. Apparently some API rate limit was exceeded, so the status couldn't be updated

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1840,7 +1800,9 @@ static Tensor& linalg_norm_out_impl(Tensor& result, const Tensor& self, optional
-      _linalg_norm_vector_out(result, self, opt_num_ord, dim_, keepdim, opt_dtype);
+      Tensor result_ = at::linalg_vector_norm(self, opt_num_ord, opt_dim, keepdim, opt_dtype);
+      resize_output(result, result_.sizes());
+      result.copy_(result_);
```
Per our offline discussion, let's add a comment here with a link to an issue for updating the out= copy behavior (to be safe castable and check the device)

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1851,6 +1813,74 @@ static Tensor& linalg_norm_out_impl(Tensor& result, const Tensor& self, optional
+
+  IntArrayRef dim = opt_dim.value_or(IntArrayRef{});
+
+  // For NumPy compatibility, we can only perform order +/-infinity reduction
```
Per offline discussion, let's be very safe in this PR and throw an error for now when dealing with no elements and a negative ord to address https://github.com/pytorch/pytorch/issues/52783. That way we can change the behavior later (if we want) without making a BC-breaking change.

We'll need to add a test for this, and it's not NumPy-compatible, but that's OK.

kurtamohler replied:
I've added an error that triggers when the input is empty and ord is negative

mruberry commented on aten/src/ATen/native/cpu/LinearAlgebraKernel.cpp:
```diff
@@ -79,8 +81,50 @@ void addr_kernel(TensorIterator &iter,
+  } else if (val.isFloatingPoint()) {
+     ord_val = val.to<double>();
+  } else {
+     AT_ERROR("linalg_vector_norm_kernel_cuda_impl expects norm to be integer or float");
```
`AT_ERROR -> TORCH_CHECK(false,`

mruberry replied:
However, should this be an internal assertion or can it be caused by a user? If it can be caused by the user, then the error message should be updated to not refer to an internal function name and the macro should be TORCH_CHECK, not AT_ERROR or TORCH_INTERNAL_ASSERT.

mruberry replied:
Similar comment as below. Also see comment about use of the word "norm."

kurtamohler replied:
There isn't any way for this error to be triggered by the user--only by a contributor. The error is only triggered if the `linalg_vector_norm_stub` call in aten/src/ATen/native/LinearAlgebra.cpp is given neither an integer nor float type `ord`. At the moment, `ord` is always converted to a double for this call.

mruberry commented on aten/src/ATen/native/cuda/LinearAlgebra.cu:
```diff
@@ -492,8 +495,70 @@ void addr_kernel_cuda(TensorIterator &iter, Scalar beta, Scalar alpha) {
+  } else if (val.isFloatingPoint()) {
+     ord_val = val.to<double>();
+  } else {
+     TORCH_CHECK(false, "linalg_vector_norm_kernel_cuda_impl expects norm to be integer or float");
```
If this can be triggered by a user then the function name should be the user-facing name "vector_norm" instead of the internal implementation name. 

The word "norm" here should match the documentation's name for the argument, too ("ord", right?)

mruberry commented on aten/src/ATen/native/cuda/LinearAlgebra.cu:
```diff
@@ -492,8 +495,70 @@ void addr_kernel_cuda(TensorIterator &iter, Scalar beta, Scalar alpha) {
+  }
+  // For complex outputs, the above kernels do not touch the imaginary values,
+  // so we must zero them out
+  if (isComplexType(iter.output().scalar_type())) {
```
@ngimel points out this is a problem for torch.norm(), too

@mruberry asks if we can address this in the future to avoid the zeroing

@ngimel says we should address this in the future

mruberry commented on aten/src/ATen/native/cuda/LinearAlgebra.cu:
```diff
@@ -492,8 +495,70 @@ void addr_kernel_cuda(TensorIterator &iter, Scalar beta, Scalar alpha) {
+#if defined(__HIPCC__)
+  if (iter.input_dtype() == kHalf) {
+    return linalg_vector_norm_kernel_cuda_impl<at::Half, float>(iter, ord);
+  } else if (iter.dtype(1) == kHalf && iter.input_dtype() == kFloat) {
```
How is `iter.dtype(1)` distinct from `iter.input_dtype()`? A comment here identifying the inputs/outputs and their positions would help, too.

kurtamohler replied:
In this case, `iter.dtype(1)` is the output's dtype. I'll change it to `iter.output_dtype()`, since that's clearer

mruberry replied:
Wait I thought that outputs had to appear before inputs in TensorIterator?

ngimel replied:
Wait, how can `iter.dtype(1)` be input dtype? TI does not work this way. `iter.output_dtype()` translates into `iter.dtype(1)` for an op with 1 output.

kurtamohler replied:
Unless I'm misunderstanding something, I think Mike is right--outputs are placed in front of inputs in `TensorIterator::operands_`. When calling `TensorIterator::input(int arg)`, `arg` is offset by the number of outputs: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/TensorIterator.h#L219

So `iter.dtype(1)` is equivalent to `iter.input_dtype()`. That can also be verified by looking at the implementation for `input_dtype()` and `dtype()`: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/TensorIterator.h#L198 and https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/TensorIterator.h#L193

This means that the above `iter.dtype(1) == kHalf && iter.input_dtype() == kFloat` condition is always false. I copied this code from `norm_kernel_cuda`, assuming that it was correct: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/ReduceNormKernel.cu#L42

But it turns out that this code is incorrect because of a commit I made. The conditional used to be `iter.dtype(1) == kHalf && iter.dtype() == kFloat`. This is almost definitely where I introduced the change that broke float16 performance for CUDA.

I'll fix this in this PR for `vector_norm`. Should we also fix it for `torch.norm` in this PR, or would it be best to do that in another PR?

Also, as side note, it turns out `TensorIterator::output_dtype()` does not exist, so I assume it's probably best to use `iter.output().scalar_type()` for readability.

ngimel replied:
The change that broke half performance was the change to `make_reduction` invocation: https://github.com/pytorch/pytorch/commit/54f0556ee453d1c4a9c6756bfcdce1e0d62867a8#diff-105d96b43a1224f639b222181fc67530f4edea19eacd33db8c57fe6273e0303cL678, `make_reduction` has to be called with out_dtype only to do the right thing (this is very fragile and should be fixed really). `make_reduction` itself, when called wrong, is doing the case, so after it nothing can be done.

kurtamohler replied:
I'm happy to fix the `make_reduction` call, but could you please give a bit more detail? I don't understand why using `out_dtype` is incorrect or how it should be fixed.

Maybe an alternative to `out_dtype` would be to create an empty result tensor directly (only if it's not defined already) and pass that into `make_reduction`? Something like this:

```
if (!result.defined()) {
  result = at::empty({0}, self.options().dtype(out_dtype));
}
auto iter = make_reduction("vector_norm", result, self_, dim, keepdim, in_dtype);
```

Is that what you'd prefer @ngimel, or something else?

kurtamohler replied:
I tried my own suggestion from above, and it doesn't work. If result is defined and its dtype is not the same as self_, then I get an error. I also get an error if result is not defined and the input is complex:

```
>>> a = torch.randn(10, dtype=torch.cfloat)
>>> result_out = torch.empty((0), dtype=torch.float)
>>> torch.linalg.vector_norm(a, 0, out=result_out)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: vector_norm: provided dtype must match dtype of result. Got Float and ComplexFloat.
>>> torch.linalg.vector_norm(a, 0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: vector_norm: provided dtype must match dtype of result. Got Float and ComplexFloat.
```

I'm not sure how else we could avoid the `out_dtype` argument for `make_reduction`

mruberry replied:
Hey @kurtamohler, @ngimel and @mruberry here. Let's talk about two things:

- First, let's fix that bug in this PR here and for torch.norm and ensure the code path is exercised properly whether that's with a test (if possible) or just manually by verifying a print statement is triggered, for example.
- Second, let's talk about make_reduction(). This is the overload that would need to be called for special half handling:

https://github.com/pytorch/pytorch/blob/c9d37675b2a44794d04c33bdaba0c40c7efabece/aten/src/ATen/native/ReduceOpsUtils.h#L196

And we would like to restore that functionality, but NOT in this PR.

kurtamohler replied:
I believe I've fixed this for `vector_norm`. But only two of these four conditions are being tested (verified with a print statement)--both of the cases that check if the output is a half precision type. The other two conditions cannot be tested at the moment, because they rely on the input and output types being different (half and float, respectively). Since this PR is throwing an error if the result tensor's type does not match the `dtype` argument (if given) or the value type of the input, then the TensorIterator's output and input dtypes will always match

mruberry commented on aten/src/ATen/native/cuda/LinearAlgebra.cu:
```diff
@@ -492,8 +495,70 @@ void addr_kernel_cuda(TensorIterator &iter, Scalar beta, Scalar alpha) {
+    // type promotion that does cast and reduction in a single kernel
+    return linalg_vector_norm_kernel_cuda_impl<at::Half, float, float>(iter, ord);
+  }
+  else if(iter.input_dtype() == kBFloat16) {
```
Are there any tests for this behavior or the previous one?

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -9074,6 +9074,17 @@
   python_module: linalg
   variants: function
 
+- func: linalg_vector_norm(Tensor self, Scalar? ord=None, int[1]? dim=None, bool keepdim=False, *, ScalarType? dtype=None) -> Tensor
```
This accepts multiple dims, right? If so, "dim" -> "dims".

mruberry commented on test/test_linalg.py:
```diff
@@ -945,7 +945,7 @@ def get_value_type(dtype):
                 self.assertEqual(result.dtype, from_dtype, msg=msg)
 
-            result_out = torch.empty((), dtype=to_dtype, device=device)
+            result_out = torch.empty((0), dtype=to_dtype, device=device)
```
What's up with this change?

kurtamohler replied:
It works fine either way, but I was getting this `resize_output` warning with size `()`, so I figured it's better to use `(0)`:

```
UserWarning: An output with one or more elements was resized since it had shape [], which does not match the required output shape [1, 1].This behavior is deprecated, and in a future PyTorch release outputs will not be resized unless they have zero elements. You can explicitly reuse an out tensor t by resizing it, inplace, to zero elements with t.resize_(0).
```

mruberry commented on test/test_linalg.py:
```diff
@@ -994,6 +994,58 @@ def get_value_type(dtype):
+        # This test compares torch.linalg.vector_norm's output with
+        # torch.linalg.norm given a flattened tensor
+        ord_vector = [0, 0.5, 1, 2, 3, inf, -0.5, -1, -2, -3, -inf, None]
+        test_cases = [
```
Separate the error cases into a different test for granularity, like "test_vector_norm_exceptions"

mruberry commented on test/test_linalg.py:
```diff
@@ -994,6 +994,58 @@ def get_value_type(dtype):
+
+        norm_dtypes = [None, torch.float, torch.double, torch.cfloat, torch.cdouble]
+
+        for (input_size, error_ords), ord, keepdim, norm_dtype in product(test_cases, ord_vector, [True, False], norm_dtypes):
```
This isn't testing the dim(s?) argument?

kurtamohler replied:
I've added dim arg cases to this test

mruberry commented on test/test_linalg.py:
```diff
@@ -1277,9 +1329,28 @@ def gen_error_message(input_size, ord, keepdim, dim=None):
 
+    # Test that linal.vector_norm gives the same result as numpy when inputs
+    # contain extreme values (inf, -inf, nan)
+    @unittest.skipIf(IS_WINDOWS, "Skipped on Windows!")
```
Why is this skipped on Windows and Mac?

mruberry commented on test/test_linalg.py:
```diff
@@ -1277,9 +1329,28 @@ def gen_error_message(input_size, ord, keepdim, dim=None):
+    # contain extreme values (inf, -inf, nan)
+    @unittest.skipIf(IS_WINDOWS, "Skipped on Windows!")
+    @unittest.skipIf(IS_MACOS, "Skipped on MacOS!")
+    @skipCUDAIfNoMagma
```
Do we really need MAGMA or LAPACK? Seems like for vector norms it's never needed

mruberry commented on test/test_linalg.py:
```diff
@@ -1277,9 +1329,28 @@ def gen_error_message(input_size, ord, keepdim, dim=None):
+    def test_vector_norm_extreme_values(self, device):
+        vector_ords = [0, 1, 2, 3, inf, -1, -2, -3, -inf]
+        vectors = []
+        for pair in itertools.product([inf, -inf, 0.0, nan, 1.0], repeat=2):
```
Just create the tensor as a fixture instead of constructing it programmatically.

mruberry commented on test/test_linalg.py:
```diff
@@ -1310,19 +1381,22 @@ def is_broken_matrix_norm_case(ord, x):
                         return True
+                if ord in ['nuc', 2, -2]:
+                    # The rest of nuc and matrix +/-2 are broken because new changes to
+                    # svd: https://github.com/pytorch/pytorch/pull/48436")
```
Let's link to the issue instead.

mruberry commented on test/test_linalg.py:
```diff
@@ -1634,6 +1708,18 @@ def gen_error_message(input_size, p, keepdim, dim=None):
                     self.assertEqual(res, expected, msg=msg)
 
+    # Test that torch.norm with p=+/-inf propagates NaN
+    @skipCUDAIfNoMagma
```
MAGMA or LAPACK shouldn't be needed for this test, either

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -431,6 +431,77 @@
+
+        Default: ``None``
+
+    dim (int, tuple of ints, list of ints, optional): If :attr:`dim` is an int,
```
We should definitely test this accepts a tuple properly, and because it can accept a tuple let's name this "dims."

kurtamohler replied:
I've added `test_vector_norm_dim_tuple_arg` in `test/test_linalg.py`

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -431,6 +431,77 @@
+vector_norm = _add_docstr(_linalg.linalg_vector_norm, r"""
+linalg.vector_norm(input, ord=None, dim=None, keepdim=False, *, dtype=None, out=None) -> Tensor
+
+Returns the vector norm of a given tensor. 
```
"Returns" -> "Computes"

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -431,6 +431,77 @@
+
+    dim (int, tuple of ints, list of ints, optional): If :attr:`dim` is an int,
+        the vector norm is calculated over the specified dimension. If
+        :attr:`dim` is a tuple of ints, the vector norm is calculated over the
```
"over the elements in the specified dimensions"

the language about "flatten[ing]" can be removed.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -419,6 +419,52 @@ def sample_inputs_linalg_norm(op_info, device, dtype, requires_grad):
+        for ord, dim, keepdim in product(ords, dim_args, [False, True]):
+            # TODO: remove this check when `max` is implemented for
+            #       float16 and bfloat16. Issue:
+            #       https://github.com/pytorch/pytorch/issues/50790
```
This issue seems resolved?

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -419,6 +419,52 @@ def sample_inputs_linalg_norm(op_info, device, dtype, requires_grad):
         return inputs
 
+def sample_inputs_linalg_vector_norm(op_info, device, dtype, requires_grad):
+    test_sizes = [
```
These generated inputs seem very thorough, which is great, does it make the grad and/or gradgradchecks take a long time to run, however?

kurtamohler replied:
You're right, this does take a very long time. I just realized that the return statement below is indented too far, so it was only adding the first test size. After fixing that, the command `python test/test_ops.py -k vector_norm` takes more than 20 minutes to complete on my machine. I'll prune down some of the test cases

kurtamohler replied:
After making some changes, `$ python test/test_ops.py -k vector_norm` takes about 70 seconds to run for me. Is this alright, or would it be better to prune it further?

mruberry replied:
70s is a long time. Are there any tests that take a particular long time? Always using (2, 2) for the 2D size sounds good. Adding an empty tensor input may be OK, too. If we want to prune further maybe the 3.5 or 2 ord can go?

mruberry replied:
@ngimel and I recommend removing 2 and 3.5 for now. Also always using (2, 2) instead of (S, S)

kurtamohler replied:
After my last update, `python test/test_ops.py -k vector_norm` now takes 35 seconds. If that's still too long, we could consider not testing both `keepdim=True` and `keepdim=False` for all cases

mruberry reviewed (commented):
Hey @kurtamohler! @ngimel and I took another look and made a few inline comments. Overall things look really good. We'll try to review the issue of identity elements with negative orders soon, too.

ngimel commented on aten/src/ATen/native/SharedReduceOps.h:
```diff
@@ -21,9 +21,26 @@
+}
+template <typename scalar_t>
+inline C10_DEVICE scalar_t min_propagate_nan(scalar_t a, scalar_t b) {
+  if (at::_isnan(a) || at::_isnan(b)) {
```
slightly simpler way is
```
scalar_t min = std::isnan(b) ? b : std::min(a,b); //std::min returns first arg if one of the args is nan
```

kurtamohler replied:
I'm getting some ROCm failures and I think (not 100% sure) that this is the culprit: https://github.com/ROCm-Developer-Tools/HIP/issues/2209

I'll add the isnan check of `a` for back in for HIP, hopefully that fixes it.

kurtamohler replied:
Confirmed, adding an `at::_isnan(a)` check for HIP fixed the failures. I added a comment with a link to the above HIP issue

kurtamohler commented:
@mruberry and @ngimel, unfortunately, I found an issue with the out variant. When the result tensor is not the same type as the default output type, the result is not correct:

```
>>> a = torch.randn(10, dtype=torch.float32)
>>> result = torch.empty((0), dtype=torch.float64)
>>> torch.linalg.vector_norm(a, 0)
tensor(10.)
>>> torch.linalg.vector_norm(a, 0, out=result)
tensor(5.3982e-315, dtype=torch.float64)
```

But using the `dtype` argument, we do get the correct result:

```
>>> torch.linalg.vector_norm(a, 0, dtype=torch.float64)
tensor(10., dtype=torch.float64)
```

`torch.norm` has the same issue:

```
>>> torch.norm(a, 0, out=result)
tensor(5.3982e-315, dtype=torch.float64)
```

Evidently, I never added tests for this. I don't know the best way to fix it yet, but one way that works is to use a temporary result tensor and then resize and copy, like this:

```
  ScalarType out_dtype = opt_dtype.has_value() ? opt_dtype.value() : toValueType(self.scalar_type());
  Tensor result_;
  auto iter = make_reduction("vector_norm", result_, self_, dim, keepdim, in_dtype, out_dtype);
  linalg_vector_norm_stub(iter.device_type(), iter, ord);

  if (result.defined()) {
    resize_output(result, result_.sizes());
    result.copy_(result_);
  } else {
    result = result_;
  }
  return result;
```

Obviously, this gives worse performance for the out variant unfortunately, but it shouldn't affect the non-out performance. I don't know how else to fix it at the moment. Is this fine for now, or should I look for a fix that avoids the resize and copy?

mruberry commented:
> @mruberry and @ngimel, unfortunately, I found an issue with the out variant. When the result tensor is not the same type as the default output type, the result is not correct:
> 
> ```
> >>> a = torch.randn(10, dtype=torch.float32)
> >>> result = torch.empty((0), dtype=torch.float64)
> >>> torch.linalg.vector_norm(a, 0)
> tensor(10.)
> >>> torch.linalg.vector_norm(a, 0, out=result)
> tensor(5.3982e-315, dtype=torch.float64)
> ```
> 
> But using the `dtype` argument, we do get the correct result:
> 
> ```
> >>> torch.linalg.vector_norm(a, 0, dtype=torch.float64)
> tensor(10., dtype=torch.float64)
> ```
> 
> `torch.norm` has the same issue:
> 
> ```
> >>> torch.norm(a, 0, out=result)
> tensor(5.3982e-315, dtype=torch.float64)
> ```
> 
> Evidently, I never added tests for this, so I will add them. I don't know the best way to fix it yet, but one way that works is to use a temporary result tensor and then resize and copy, like this:
> 
> ```
>   ScalarType out_dtype = opt_dtype.has_value() ? opt_dtype.value() : toValueType(self.scalar_type());
>   Tensor result_;
>   auto iter = make_reduction("vector_norm", result_, self_, dim, keepdim, in_dtype, out_dtype);
>   linalg_vector_norm_stub(iter.device_type(), iter, ord);
> 
>   if (result.defined()) {
>     resize_output(result, result_.sizes());
>     result.copy_(result_);
>   } else {
>     result = result_;
>   }
>   return result;
> ```
> 
> Obviously, this gives worse performance for the out variant unfortunately, but it shouldn't affect the non-out performance. I don't know how else to fix it at the moment. Is this fine for now, or should I look for a fix that avoids the resize and copy?

What would you think about restricting out= to simply be the "expected" or "correct" dtype for now? This is technically BC-breaking, but since it didn't work we can't actually be breaking anyone. So complex input would require a float out=, for example.

In the future we can revisit being more liberal about out= casting.

What are your thoughts?

kurtamohler commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1939,6 +1971,28 @@ def _make_tensor_helper(shape, low=None, high=None):
+                        device_type='cpu',
+                        dtypes=[torch.float16, torch.bfloat16]),
+               # TODO: remove this when test_variant_consistency_jit_cuda
+               #       failures are fixed for half and bfloat16
```
I discovered another issue here when I enabled 2-D inputs in `sample_inputs_linalg_vector_norm`. For CUDA half precision types, 2-D inputs seem to be causing the double gradient to be NaN. At least I think that's what's happening, not entirely sure yet

kurtamohler replied:
The issue seems to come from here: https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_jit.py#L100

Within one of the calls to this function during `test_variant_consistency_jit_linalg_vector_norm_cuda_float16`, the magnitudes of the numbers in this multiplication are too large for float16, so the result becomes infinite:

```
l1_test:
2480.0
allSum(grads_test):
-187.875
allSum(grads_test) * l1_test:
-inf
```

One way to get around this issue is to just make the input size smaller, so I'll do that unless there's a better solution

kurtamohler replied:
I had to decrease the 2-D input size for torch.half to `(2, 2)` in order to get the test to pass

mruberry replied:
That sounds reasonable. I also have a PR that will limit this test to only run in float32. The signal we get from other dtypes isn't very good and these workaround are a pain. It hasn't landed yet, however.

kurtamohler commented:
> What would you think about restricting out= to simply be the "expected" or "correct" dtype for now?

I've updated to throw an error if `out.dtype` does not match the `dtype` argument (if it's given) or the value type of the input (if `dtype` arg is not given).

I added a test to make sure the error is thrown correctly, `test_vector_norm_out_dtype_error` in `test/test_linalg.py`

mruberry commented:
Hey @kurtamohler, just checking in. I made some comments where I saw you had questions. Keep them coming and be sure to ping @ngimel and me when this is ready for another review!

kurtamohler commented on test/test_linalg.py:
```diff
@@ -1238,6 +1240,136 @@ def get_value_type(dtype):
+    # Test that linalg.vector_norm throws an error if the out tensor's dtype
+    # does not match the expected output dtype
+    @dtypes(torch.float, torch.double, torch.cfloat, torch.cdouble, torch.bfloat16, torch.float16)
+    def test_vector_norm_out_dtype_error(self, device, dtype):
```
Note: I added this test recently as a follow up to https://github.com/pytorch/pytorch/pull/51099#issuecomment-791046586

It checks that an error is thrown if `result.dtype` does not match the expected dtype (`self.dtype` or `dtype` arg). It also ensures that an error is thrown if `result` is complex and the `dtype` arg is not complex

kurtamohler commented:
@mruberry, thanks for your comments! I believe I've updated the PR to align with everything we've discussed so far, so it's ready for another review

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1851,6 +1815,89 @@ static Tensor& linalg_norm_out_impl(Tensor& result, const Tensor& self, optional
+        }
+      }
+      TORCH_CHECK(has_identity,
+        "vector_norm cannot compute the infinity norm on an empty tensor ",
```
this error message seems misleading because the infinity norm can be computed on an empty tensor, it just can't be computed on a dimension of size zero

kurtamohler replied:
Updated to say "empty dimension" instead of "empty tensor"

mruberry commented on aten/src/ATen/native/cpu/LinearAlgebraKernel.cpp:
```diff
@@ -79,8 +81,48 @@ void addr_kernel(TensorIterator &iter,
+  if (ord.isFloatingPoint()) {
+     ord_val = ord.to<double>();
+  } else {
+     AT_ERROR("linalg_vector_norm_kernel_cuda_impl expects ord to be float");
```
`TORCH_CHECK(false, ...`

`AT_ERROR` is deprecated.

mruberry replied:
If this error can be triggered by user input then the name should be the user-facing name, too:

"linalg.vector_norm expects ..."

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1851,6 +1815,89 @@ static Tensor& linalg_norm_out_impl(Tensor& result, const Tensor& self, optional
+  auto ord = opt_ord.value_or(2).toDouble();
+
+  TORCH_CHECK(self.device().type() == DeviceType::CPU || self.device().type() == DeviceType::CUDA,
+              "vector_norm only supports CPU and CUDA device types, but got: ", self.device().type());
```
vector_norm -> linalg.vector_norm

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1851,6 +1815,89 @@ static Tensor& linalg_norm_out_impl(Tensor& result, const Tensor& self, optional
+  // Casting a large integer to a double will introduce some error, but for
+  // practical purposes, it won't matter since a large order will usually
+  // give an infinite result
+  auto ord = opt_ord.value_or(2).toDouble();
```
Follow-up question, what happens if ord was complex?

kurtamohler replied:
```
>>> torch.linalg.vector_norm(a, ord=1j)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: value cannot be converted to type double without overflow: (0,1)
```

mruberry reviewed (commented):
Hey @kurtamohler! @ngimel and I were reviewing this and it looks good so far. We have to stop for today, however, and will try to pickup again tomorrow.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -420,6 +420,42 @@ def sample_inputs_linalg_norm(op_info, device, dtype, requires_grad):
+    ords = (None, 0, 0.5, 1, 2, 3.5, inf, -3.5, -inf)
+    inputs = []
+
+    def get_all_unique_dim_args(ndim):
```
For simplicity let's write the test cases as tuples (shape, args) and just handwrite some interesting set of dims to inspect instead of programmatically generating all of them. There is only a 1D and 2D tensor, anyway, so the interesting arguments to dim are few.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -420,6 +420,42 @@ def sample_inputs_linalg_norm(op_info, device, dtype, requires_grad):
+    for test_size in test_sizes:
+        dim_args = get_all_unique_dim_args(len(test_size))
+
+        for ord, dim, keepdim in product(ords, dim_args, [False, True]):
```
See above. Let's just handwrite the few interesting cases for now.

mruberry reviewed (approved):
Hey @kurtamohler! @ngimel and I made a few comments about small tweaks. Just ping me when this PR is ready to be merged.

It's awesome to have torch.linalg.vector_norm()! And exciting to have a torch.linalg package that's more than just a clone of np.linalg (cc @IvanYashchuk). Let's sync offline on how we want to follow-up.

kurtamohler commented:
@mruberry, I think it's ready, unless we get any CI failures or if 35 seconds is still too long for the Ops tests
